kv/txn_coord_sender.go

```diff
@@ -396,14 +396,26 @@ func (tc *TxnCoordSender) sendOne(ctx context.Context, call proto.Call) {
 				return
 			}
 			tc.Lock()
-			if txnMeta := tc.txns[id]; id != "" && txnMeta != nil {
+			txnMeta, metaOK := tc.txns[id]
+			if id != "" && metaOK {
 				args.Intents = txnMeta.intents()
 			}
 			tc.Unlock()
-			// If there aren't any intents, then there's factually no
-			// transaction to end. Read-only txns have all of their state in
-			// the client.
-			if len(args.Intents) == 0 {
+
+			if !metaOK {
+				// If we don't have the transaction, then this must be a retry
+				// by the client. We can no longer reconstruct a correct
+				// request so we must fail.
+				//
+				// TODO(bdarnell): if we had a GetTransactionStatus API then
+				// we could lookup the transaction and return either nil or
+				// TransactionAbortedError instead of this ambivalent error.
+				call.Reply.Header().SetGoError(util.Errorf(
```

Comment:
this is the cause of the regression in https://github.com/cockroachdb/cockroach/issues/2145. I believe there's a missing `return` after this line.